### PR TITLE
SQL: LOOKUP function to perform a value replacement #3555 + LOOKUP function doesn't support tags, entity.tags, metric.tags fields #3769

### PR DESCRIPTION
**Pull request by RawrLion**
added tests for #3555

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        assertSqlQueryRows("LOOKUP in WHERE gives wrong result", expectedRows, sqlQuery);
+    }
+
+    /**
+     * #3555
+     */
+    @Test
+    public void testLookupInCast() {
+        String sqlQuery = String.format(
+                "SELECT CAST(LOOKUP('table', t1.text)) " +
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: Write explicit type to cast (it is more verbose)

File: src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Map;
+
+public class ReplacementTableMethod extends BaseMethod{
+    private static final String PATH = "/entities/lookup";
+    private static final WebTarget resource = httpRootResource.path(PATH);
+
+    public static void create(ReplacementTable table) throws UnsupportedEncodingException {
+        Form form = new Form();
+        form.param("lookupName", table.getName());
+
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: What is a  purpose of linebreaks after each line?

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+
+        SeriesMethod.insertSeriesCheck(seriesList);
+    }
+
+    @DataProvider(name = "lookupWithTagsTestProvider")
+    public Object[][] provideTestsDataForLookupWithTagsTest() {
+        return new Object[][]{
+                {
+                        "1", "negative"
+                },
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: May be joining it into single line will be more compact?

File: src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.axibase.tsd.api.model.replacementtable.ReplacementTable;
+
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.Response;
+import java.io.UnsupportedEncodingException;
+import java.util.Map;
+
+public class ReplacementTableMethod extends BaseMethod{
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: format code

File: src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        Form form = new Form();
+        form.param("lookupName", table.getName());
+        form.param("items", squashMapIntoString(table.getMap()));
+        form.param("oldName", "");
+        form.param("save", "Save");
+
+        Response response = resource
+                .request()
+                .post(Entity.form(form));
+        response.bufferEntity();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: should we get response body and also check that we correctly save replacement table?

File: src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.Map;
+
+public class ReplacementTableMethod extends BaseMethod{
+    private static final String PATH = "/entities/lookup";
+    private static final WebTarget resource = httpRootResource.path(PATH);
+
+    public static void create(ReplacementTable table) throws UnsupportedEncodingException {
+        Form form = new Form();
+        form.param("lookupName", table.getName());
+        form.param("items", squashMapIntoString(table.getMap()));
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: Can we use something like MultiValueMap here?

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.axibase.tsd.api.util.Util.TestNames.entity;
+import static com.axibase.tsd.api.util.Util.TestNames.metric;
+
+public class SqlLookupFunctionTest extends SqlTest {
+    private static final String TEST_METRIC_NAME = metric();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: rename constants to a more appropriate ones.

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+    private static final String TEST_METRIC_NAME1 = metric();
+    private static final String TEST_ENTITY_NAME = entity();
+    private static final String TEST_ENTITY_NAME1 = entity();
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: Move it into `prepareData` method

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+
+        assertSqlQueryRows("LOOKUP gives wrong result, " +
+                "using Replacement Table with parameters: " + table1.toString(), expectedRows, sqlQuery);
+    }
+
+    /**
+     * #3555
+     */
+    @Test
+    public void testDoubleLookupFromText() {
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: testLookupComposition?

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+                {"-3"},
+                {"null"},
+                {"null"},
+                {"PI"},
+                {"PI"},
+                {"3.14"},
+                {"null"}
+        };
+
+        assertSqlQueryRows("LOOKUP in LOOKUP gives wrong result, " +
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: LOOKUP COMPOSITION ...

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+                "using Replacement Table with parameters: " + table1.toString(), expectedRows, sqlQuery);
+    }
+
+    /**
+     * #3555
+     */
+    @Test
+    public void testLookupCaseSensitiveness() {
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: Sensitivity?

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+            series.setTags(tags);
+            series.addData(new Sample("2016-06-03T09:20:00.000Z", "1"));
+            seriesList.add(series);
+        }
+
+
+        SeriesMethod.insertSeriesCheck(seriesList);
+    }
+
+    @DataProvider(name = "lookupWithTagsTestProvider")
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: Move it closer to tests using it

File: src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+     */
+    @Test
+    public void testDoubleLookupFromText() {
+        String sqlQuery = String.format(
+                "SELECT LOOKUP('tableForLookupTest1', LOOKUP('tableForLookupTest1', t1.text)) " +
+                        "FROM '%s' t1",
+                TEST_METRIC_NAME
+        );
+
+        String[][] expectedRows = { // looking twice in Replacement Table
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/sql/operator/SqlLookupFunctionTest.java]: Is there any reason to place comments so?

File: src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java [hunk 14]
```diff
@@ -1,34 +1,53 @@
...
 
         Response response = resource
                 .request()
                 .post(Entity.form(form));
         response.bufferEntity();
+
+        return response;
+    }
+
+    public static void createCheck(ReplacementTable table) throws UnsupportedEncodingException {
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/replacementtable/ReplacementTableMethod.java]: WOW, Such exception

**Pull request merged**